### PR TITLE
fix(Core/Player): learnSkillRewardedSpells AFTER ALL skills have been loaded

### LINKED ISSUE #6670: [Tauren] Cultivation Racial not being applied after log off and back on
#### WHAT CLIENT DO YOU PLAY ON?

- [ ] enGB
- [X] enUS
- [ ] other (specify)

##### FACTION

- [ ] Alliance
- [X] Horde

##### CONTENT PHASE:

- [X] Generic
- [ ] 1-19
- [ ] 20-29
- [ ] 30-39
- [ ] 40-49
- [ ] 50-59

##### CURRENT BEHAVIOUR:

I am currently playing a Tauren druid "Keir" level 13. When I started yesterday I trained in herbalism and saw that I got the extra 15 point due to my racial. Putting me at 16 and rank 1 herbalism completed at a skill of 90. I leveled it up a bit as I was leveling yesterday and logged off to head in to work.

Once I logged back in again I went to pick a herb "earthroot that has a skill of 15 and was promptly unable to pick the herb due to not having enough skill.

With that I unlearned the skill trained again, attempted to pick the level 15 herb and was successful gaining 1 point in my skill to 16 with a max skill being able to be raised to 90. I logged out and then back in and checked my skill and it was at a skill of 2 with a max skill of 75.

It is as if the Tauren race bonus is not being applied..

Confirmed the issue on a local AC server: https://youtu.be/MFsoQpU8vg4
AzerothCore rev. c8a4b8190cb6+ 2021-06-30 22:47:41 +0000 (master branch)

##### EXPECTED BLIZZLIKE BEHAVIOUR:

Tauren race gets a 15 point skill bump for herbalism. 15 skill to start and a max skill for 315 at 60. I do not have any video proving it but I have seen it discussed many times... Here is one example
https://www.reddit.com/r/classicwow/comments/b74r82/what_is_the_purpose_of_the_tauren_racial/

##### SOURCE:

https://wowwiki-archive.fandom.com/wiki/Cultivation
https://wotlk.evowow.com/?spell=20552

##### STEPS TO REPRODUCE THE PROBLEM:

Originally reported https://github.com/chromiecraft/chromiecraft/issues/1046

##### AC HASH/COMMIT:

https://github.com/chromiecraft/azerothcore-wotlk/commit/cecbd604d3f57312165cc53f8ed40feae51ee45f

##### OPERATING SYSTEM:

Ubuntu 20.04

##### MODULES:

- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-chromie-xp](https://github.com/azerothcore/mod-chromie-xp)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-eluna-lua-engine](https://github.com/azerothcore/mod-eluna-lua-engine)
- [mod-ip-tracker](https://github.com/azerothcore/mod-ip-tracker)
- [mod-low-level-arena](https://github.com/azerothcore/mod-low-level-arena)
- [mod-multi-client-check](https://github.com/azerothcore/mod-multi-client-check)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)
- [mod-pvpstats-announcer](https://github.com/azerothcore/mod-pvpstats-announcer)
- [mod-queue-list-cache](https://github.com/azerothcore/mod-queue-list-cache)
- [mod-server-auto-shutdown](https://github.com/azerothcore/mod-server-auto-shutdown)
- [lua-CarbonCopy](https://github.com/55Honey/Acore_CarbonCopy)
- [lua-LevelUpReward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-send-and-bind](https://github.com/55Honey/Acore_SendAndBind)
- [lua-Zonecheck](https://github.com/55Honey/acore_Zonecheck)

##### OTHER CUSTOMIZATIONS:

None.

##### SERVER:

ChromieCraft

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/99600468-tauren-cultivation-racial-not-being-applied-after-log-off-and-back-on?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by necropola**
## Changes Proposed:
- Call `learnSkillRewardedSpell` **after all** skills have been loaded  (copied from TC-3.3.5) 
- This guarantees that the **Herbalism (182)** skill is loaded when the aura spell **Cultivation (20552)** (which grants a permanent +15 skill bonus) is learned from the **Tauren Racial (124)** skill.    

## Issues Addressed:
- Closes #6670
- Closes https://github.com/chromiecraft/chromiecraft/issues/1046

## SOURCE:
- The commit that added this to TC-3.3.5, but for a different reason: https://github.com/TrinityCore/TrinityCore/commit/9ee820d83d2bc3a2fb7c88047874c6420b116799

## Tests Performed:
- Server successfully compiled and started
- Tested in game

## How to Test the Changes:
- Learn **Herbalism** on a **Tauren**
- Logoff + Logon
- Observe that the racial bonus to **Herbalism** is still applied

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Review by FrancescoBorzi — APPROVED** [review]

**Review by Kitzunu — APPROVED** [review]
Good

**Comment by necropola** [comment]
@Kitzunu When I made the PR, I just took the lines from current TC-3.3.5, because I thought that they might fix the issue. And they did. Meanwhile I did some research and found the TC commit that actually added exactly those lines: https://github.com/TrinityCore/TrinityCore/commit/9ee820d83d2bc3a2fb7c88047874c6420b116799

The interesting part is that the commit actually addressed something else. And then I remembered those pink SQL warning messages on my local AC that recently started to appear whenever I logged a character on my lolcal AC server. They are all gone now. Nice side (or rather main) effect, I'd say.

I'm currently playing around with my many characters on my local AC server to see whether my PR has some weird side effects. But so far, eveyrthing looks good.

File: src/server/game/Entities/Player/Player.cpp [hunk 1]
```diff
@@ -26085,6 +26086,12 @@ void Player::_LoadSkills(PreparedQueryResult result)
         } while (result->NextRow());
     }
 
+    // Learn skill rewarded spells after all skills have been loaded to prevent learning a skill from them before its loaded with proper value from DB
+    for (auto& skill : loadedSkillValues)
+    {
+        learnSkillRewardedSpells(skill.first, skill.second);
```
Comment (claudiodfc) [inline src/server/game/Entities/Player/Player.cpp:26092]: What exactly is SkillRewardedSpells?

Comment (necropola) [reply src/server/game/Entities/Player/Player.cpp:26092]: Something like this: https://wowgaming.altervista.org/aowow/?skill=124
**Spells**, e. g.  **Cultivation (20552)**,  which are automatically learned when getting a **skill**,  e. g. **Tauren Racial (124)**

Comment (claudiodfc) [reply src/server/game/Entities/Player/Player.cpp:26092]: I really hate the distinction that they make between skills and spells. It's basically a passive spell which is a combination of a class + race. Thanks for clarifying though!

**Comment by necropola** [comment]
###  Test resulkts from my local AC server with PR applied
To make things a bit easier I allowed 11 Primary professions in `worldserver.conf`

- Level 39 Blood Elf Mage
  -  .reset spells + logoff + logon (default spells and skills are applied)
  - Learned all class spells from trainer
  - 30 Talent points spent
  - Learned all 11 professions with all ranks and all available recipes from trainer
     (used  `.setskill` to level up until max)
  - Learned all 3 Secondary Skills with all ranks and all available recipes from trainer
     (used  `.setskill` to level up until max
  - Learned Apprentice Riding 75/75
  - Verified that everything incl. racial profession bonus (Enchanting +10) stays when relogging
  - Killed some stuff
 
- Level 39 Draenei Mage
  -  .reset spells + logoff + logon (default spells and skills are applied)
  - Learned all class spells from trainer
  - 30 Talent points spent
  - Learned all 11 professions with all ranks and all available recipes from trainer
     (used  `.setskill` to level up until max)
  - Learned all 3 Secondary Skills with all ranks and all available recipes from trainer
     (used  `.setskill` to level up until max
  - Learned Apprentice Riding 75/75
  - Verified that everything incl. racial profession bonus (Jewelcrafting +5) stays when relogging
  - Verified that Gift of Naaru scales with level (incl. tooltip)
  - Killed some stuff
 
 - Level 39 Undead Warlock
   -  .reset spells + logoff + logon (default spells and skills are applied)
   - Learned all class spells from trainer
   - 30 Talent points spent
   - Learned all Tailoring + Enchanting with all ranks and all available recipes from trainer
      (used  `.setskill` to level up until max)
   - Learned all 3 Secondary Skills with all ranks and all available recipes from trainer
      (used  `.setskill` to level up until max
   - Learned Apprentice Riding 75/75 through Warlock Trainer
   - Verified that everything stays when relogging
   - Killed some stuff

- Level 39 Tauren Druid
   - Just verified that the racial bonus for Herbalism stays when relogging

- Level 39 Gnome Warrior (created while PR was applied)
   - Just verified that racial bonus for Engineering stays when relogging

**Pull request merged**